R/fit_trait_distributions.R

```diff
@@ -0,0 +1,130 @@
+#' in long format
+#' @param distribution_type the type of statistical distribution to use.
+#' Character or named list. Currently accepts "normal","lognormal", and "beta".
+#' @note Distribution type should be a single character (e.g. "normal")
+#' or a named list (e.g. list(height="normal",mass="lognormal"))
+#' @description
+#'
+#' @return a tibble
+#'
+#' @importFrom stats var
+#' @importFrom e1071 skewness kurtosis
+#' @importFrom magrittr %>%
+#' @importFrom dplyr slice_sample group_by summarise
+#' summarize select group_by_at
+#' @importFrom purrr map_dfr
+#' @importFrom stats var
+#' @importFrom data.table :=
+#' @importFrom fitdistrplus fitdist
+
+#' @examples
+#' data(community)
+#' data(trait)
+#' imputed_traits <-trait_impute(comm = community, traits = trait,
+#'                  scale_hierarchy = c("Site", "PlotID"),
+#'                  taxon_col = "Taxon", value_col = "Value",
+#'                  trait_col = "Trait", abundance_col = "Cover")
+#' fitted_distributions <- fit_trait_distributions(
+#' imputed_traits = imputed_traits,
+#' distribution_type = "normal")
+#'
+#' @export
+fit_trait_distributions <- function(imputed_traits,
+                                    distribution_type = "normal") {
+
+  #Pull useful information from imputed traits object
+  value_col <- attributes(imputed_traits)$attrib$value_col
+  trait_col <- attributes(imputed_traits)$attrib$trait_col
+  taxon_col <- attributes(imputed_traits)$attrib$taxon_col
+  abundance_col <- attributes(imputed_traits)$attrib$abundance_col
+  scale_hierarchy <- attributes(imputed_traits)$attrib$scale_hierarchy
+
+  #If only a single type of distribution was supplied, apply it to all traits
+  if (length(distribution_type) == 1) {
+    distributions <- unique(as.data.frame(imputed_traits)[, trait_col])
+    dist_list <- replicate(n = length(distributions), expr = distribution_type)
+    names(dist_list) <- distributions
+    distribution_type <- as.list(dist_list)
+    rm(dist_list, distributions)
+  } #if statement
+
+  #Data checks
+
+  #check distribution type is supported
+  if (any(!unique(distribution_type) %in% c("beta", "normal", "lognormal"))) {
+    x <- distribution_type[!distribution_type %in%
+                            c("beta", "normal", "lognormal")]
+    stop(glue("distribution type '{x}' not supported "))
+  }
+
+
+  #Beta checks
+  if ("beta" %in% distribution_type) {
+
+    #make sure every species x hierarchy combination has at least 2 data points
+    beta_counts <- imputed_traits %>%
+      filter(.data[[trait_col]] ==
+               names(distribution_type)[distribution_type == "beta"]) %>%
+      group_by_at(c(as.character(scale_hierarchy), trait_col, taxon_col)) %>%
+      summarise(n = n(), .groups = "keep")
+
+    if (any(beta_counts$n < 2)) {
+    stop("Fitting a beta distrbution requires 2+ points per distribution.
+         We suggest re-imputing traits with a minimum of (at least) 2 leaves.")
```

Comment:
side note - would be good if the coverage plot had some way to show the depth of coverage Thought about setting alpha, but not sure how interpretable that would be.

Reply:
Good point.  I was imagining the situation where the imputation could draw from other sites, but if there was only one observation of a species across the whole dataset, the issue would remain.   Since this issue impacts both parametric and nonparametric bs, perhaps this could be addressed with the trait imputation?  Maybe if a minimum number of leaves is specified ,and that minimum cannot be met by drawing from other levels of the hierarchy, the species is removed?